sherpa/include/sherpa/astro/xspec_extension.hh

```diff
@@ -297,6 +297,13 @@ PyObject* xspecmodelfct( PyObject* self, PyObject* args )
           XSpecFunc( &fear[0], &npts, &pars[0], &ifl,
                      &result[0], &error[0] );
 
+	} catch(std::exception& exc) {
+
+          std::ostringstream err;
+          err << "XSPEC model evaluation failed: " << exc.what();
+          PyErr_SetString( PyExc_ValueError, err.str().c_str() );
+          return NULL;
+
```

Comment:
Interesting. Any way we can test this/these?

Reply:
The only way I can think of would be to build - as part of the test - an xspec model which we can trigger a failure in certain conditions (e.g. have a parameter as 0 or 1 and if 1 then it throws an exception). Otherwise we could try and look through xspec models to see if any do raise an exception, and then see of there's a way to trigger that condition in the test (I'm thinking of changing the path to the data files, since a model may fail then). However, that's not robust (since it depends on internals of the XSPEC model library),

Reply:
From looking at the commit message I see that I added this because it helped diagnose a problem with a XSPEC user model!

Reply:
Summary of the following: it's not obvious that this condition can actually be triggered, but there is no explicit statement on this from the XSPEC docs (so we should probably ask them).

Hmmm, I think my comment about a user model must have referred to my work on building XSPEC user models which (currently) does not use this template code - i.e. I saw the error and had to add a similar throw/catch block in that code, so hence added it here also just in case. My attempts to trigger a similar situation with XSPEC models has been less fruitful. I looked through the models for ones with an obvious raise and found the laor model (amongst others). I can trigger the condition, but it's obviously getting caught/handled within the XSPEC interface itself. As an example, using CIAO (but the same behavior is seen on this branch with XSPEC 12.9.1p):

a) default set up

```
% sherpa
-----------------------------------------------------
Welcome to Sherpa: CXC's Modeling and Fitting Package
-----------------------------------------------------
CIAO 4.9 Sherpa version 1 Friday, December 2, 2016

sherpa-1> from sherpa.astro import xspec
sherpa-2> xspec.get_xsversion()
          '12.9.0o'
sherpa-3> xspec.set_xschatter(50)
sherpa-4> xslaor.mdl()
          <XSlaor model instance 'xslaor.mdl'>
sherpa-5> mdl([1.1, 1.11, 1.12])
          array([ 0.53513175,  0.46486822,  0.        ])
```

b) trigger a failure in the laor model

The model loads in data files which it gets from XSPEC_MDATA_DIR env var, if set, otherwise it is based on HEADAS. So setting XSPEC_MDATA_DIR to an invalid path triggers the error, which is converted to a message printed to stderr (I believe), and then exits. Our code makes sure that the return array in this case is all 0's.

```
% export XSPEC_MDATA_DIR /dev/null/
% sherpa
-----------------------------------------------------
Welcome to Sherpa: CXC's Modeling and Fitting Package
-----------------------------------------------------
CIAO 4.9 Sherpa version 1 Friday, December 2, 2016

sherpa-1> from sherpa.astro import xspec
sherpa-2> xspec.get_xsversion()
          '12.9.0o'
sherpa-3> xspec.set_xschatter(50)
sherpa-4> xslaor.mdl()
          <XSlaor model instance 'xslaor.mdl'>
sherpa-5> mdl([1.1, 1.11, 1.12])

***XSPEC Error: In RDLAOR, FITSIO error = 104
Failed to open /dev/null/ari.mod
          array([ 0.,  0.,  0.])
```

Reply:
A discussion with @xspector says that XSPEC aims to catch all the errors, so this particular check is likely not that useful.

Reply:
I am confused as to why the exception is not bubbling up. Shouldn't a `ValueError` Python exception be raised if any C++ exception is thrown by the wrapper xspec function? Unless I am missing something that is not the case. Or am I missing something?

Reply:
I believe that there is some "hidden" XSPEC wrapper code that lies between the implementation (that is, the function that is pointed to in the `model.dat` file) and the function name we call, and this catches C++ exceptions (or at least any of the Yellow/RedAlert exceptions that they define). So we never see the exception. Or I don't understand C++ exceptions well enough, which is a perhaps-more-realistic scenario.

The error message I show above appears to come from this bit of code from `XSFunctions/xslaor.cxx`:

```
   std::ostringstream errMsg;
   errMsg << "In RDLAOR, FITSIO error = ";
   fitsfile *fptr=0;

   // Open the ari.mod file
   fits_open_file(&fptr,cFileName, 0, &status);
   if (status)
   {
      errMsg << status << "\nFailed to open " << modFileName << "\n";
      throw YellowAlert(errMsg.str());
   }
```

A quick look at `XSUtil/Error/Error.cxx` has

```
YellowAlert::YellowAlert()
{
}

YellowAlert::YellowAlert (const string& message)
{
  *IosHolder::errHolder() << "\n***XSPEC Error: " << message << std::flush;
}
```

I'm not sure if this helps you.

Reply:
Here's the info I got from Craig Gordon when I asked if they had a policy on exception handling:

> I can't recall any cases where exceptions currently propagate entirely out of the model functions library, but we've never established a rule guaranteeing that that will always be the case.  So unless maintaining those try/catch blocks is problematic, I recommend keeping them.  If they're at a high enough level of the code, I wouldn't think it would affect performance.
>
> Just for your future reference, all Xspec exceptions inherit from either YellowAlert or RedAlert classes (defined in src/XSUtil/Error/Error.h).  RedAlerts basically perform the role of abort() calls: something so bad has happened that Xspec should terminate immediately.  Therefore in our high level try/catch blocks, we often just catch YellowAlert& while letting RedAlerts go uncaught to end the program.  Of course if there is something you want to do even for RedAlerts, you would do "catch (...)" as in your current code.

Reply:
> Our code makes sure that the return array in this case is all 0's.

I probably misunderstood this line in one of your previous comments. I thought you were claiming that our catch code was responsible for returning a zero array, so that it was called. If that's not the case then I am not confused anymore, because we have enough evidence to believe that the catch code is never ever called. Right?

Reply:
Our code zeroes out the return array before calling the XSPEC model. So if the XSPEC model doesn't do anything (presumably because of an error condition) then we will return all 0's. This was done because we used to just send in an array of uninitialized memory, which meant that if there was an error we would get random values out.

If the XSPEC model has changed some of the values before the error then we'll see those changes, but I don't think it's worth worrying about that case.

Does this help you? At present I think the catch code I added could be removed if you feel it is potentially unsafe. If we did remove it we wouldn't be in a worse case than we are know (i.e. in the released code), and we can worry about other PRs.